### DISCUSSION TITLE
z-projection for volume scans

### DISCUSSION THREAD

**Discussion by peteratBHVI** [marksgraham/OCT-Converter#23] [General]
Hi Mark, 

great work! very useful to me! Just a question, after the extraction of the images I use the free IOWA segmentation tool and it spits out a projection. It looks like a B&W photo of the retina but is generated from the images in the .fda file. Now is that already within the .fda or does it require some kind of processing? would you know? 

![images_Sequence_Proj_Iowa](https://user-images.githubusercontent.com/37785427/115098040-1c4f1e00-9f71-11eb-96e8-020223c80082.png)

The discontinuation of blood vessels indicate transversal eye movements during the scan. 
 
Best, Peter

**Comment by marksgraham** [comment]
Hi Peter,

I don't know if there is a projection stored in the .fda itself, but it is actually pretty easy to generate these projections - you take the 3D image data and sum (or average) along the z-plane to create a 2D projection image.

I can take a look at adding an option to produce these in the code - may be able to take a look sometime this week. In the meantime should be pretty easy to get these yourself by manipulating the `volume` that is returned from `oct_volume = fds.read_oct_volume() `

Best,
Mark

**Reply by peteratBHVI** [reply]
Hi Mark, 

that is awesome, so you say if I take each image and sum along the 3D depth axis (your output image height) and then assemble it image by image I should get the same projection. 
Would be nice to find that option in your code. I think I will be able to create that from your output images too. 

This projection actually helps me to assess the quality of the scan and average across scans for the same participant. 

Thanks again! 
Peter

**Comment by peteratBHVI** [comment]
Hi Mark, 
this did it for me, though the images are a bit dull. Feel free to uses snippets. 

import numpy as np
import os
import glob
from PIL import Image, ImageOps


fp = 'your_OCT_file_path' # sample: images subfolder == images, images file names == OCTid_slice_imageNR.png
arr_new = []
images_fn = sorted(glob.glob(os.path.join(fp, 'images') + '\\*_slice_*.png'),
                   key=lambda x: int(x.split('\\')[-1].split('_')[-1][0:-4])
                  )
for image_fn in images_fn:
    arr_mean_z = np.sum(Image.open(image_fn), axis=0)
    arr_new.append(arr_mean_z)
# normalize arr_all, enhance contrast and scale to grey image 
arr_all = ((np.array(arr_new)/ np.max(arr_new) )**5)*255
# auto contrast, convert to grey image 
im_out = ImageOps.autocontrast(Image.fromarray(arr_all).convert('L'), cutoff=0, ignore=None)

Best, 
Peter

**Reply by peteratBHVI** [reply]
from PIL import Image, ImageOps, ImageEnhance
 # with this at the end does the job ok. 
    im_out = ImageEnhance.Contrast(im_out)
    im_out = im_out.enhance(3.0)

cheers, 
Peter

**Comment by marksgraham** [comment]
I've added the ability to extract the projection as a np array with the `read_projection` method of `OCTVolume` or save it as an image file with `save_projection`. I agree the projections look a bit washed-out if you save directly but I'm reluctant to manipulate them too much; if users want to adust them they can use the `read_projection` method then do whatever they want with them :)

**Reply by peteratBHVI** [reply]
Legend! and I agree, having raw data enables the user to manipulate it by themselves! 
Thanks heaps!
